pkg/bpf/map.go

```diff
@@ -439,25 +496,73 @@ func (m *Map) Lookup(key MapKey) (MapValue, error) {
 }
 
 func (m *Map) Update(key MapKey, value MapValue) error {
+	var err error
+
 	m.lock.Lock()
 	defer m.lock.Unlock()
 
-	if err := m.Open(); err != nil {
+	defer func() {
+		if m.cache == nil {
+			return
+		}
+
+		desiredAction := OK
+		if err != nil {
+			desiredAction = Insert
+		}
+
+		m.cache[key.String()] = &cacheEntry{
+			Key:           key,
+			Value:         value,
+			DesiredAction: desiredAction,
+			LastError:     err,
+		}
+	}()
+
+	if err = m.Open(); err != nil {
 		return err
 	}
 
-	return UpdateElement(m.fd, key.GetKeyPtr(), value.GetValuePtr(), 0)
+	err = UpdateElement(m.fd, key.GetKeyPtr(), value.GetValuePtr(), 0)
+	return err
+}
+
+func (m *Map) deleteCacheEntry(key MapKey, err error) {
+	if m.cache == nil {
+		return
+	}
+
+	k := key.String()
+	if err == nil {
+		delete(m.cache, k)
+	} else {
+		entry, ok := m.cache[k]
+		if !ok {
+			m.cache[k] = &cacheEntry{
+				Key: key,
+			}
+			entry = m.cache[k]
+		}
+
+		entry.DesiredAction = Delete
+		entry.LastError = err
+	}
 }
 
 func (m *Map) Delete(key MapKey) error {
+	var err error
+
 	m.lock.Lock()
 	defer m.lock.Unlock()
 
-	if err := m.Open(); err != nil {
+	defer m.deleteCacheEntry(key, err)
+
+	if err = m.Open(); err != nil {
 		return err
 	}
 
-	return DeleteElement(m.fd, key.GetKeyPtr())
+	err = DeleteElement(m.fd, key.GetKeyPtr())
```

Comment:
Does this return an error if the element does not exist? Wondering how this affects the flow of `m.deleteCacheEntry` if the key does not exist, as the cache gets populated with a cacheEntry. I guess the error would give the needed information, so this may be a non-issue.

Reply:
Fixed